WordPress/src/main/java/org/wordpress/android/ui/activitylog/ActivityLogListFragment.kt

```diff
@@ -0,0 +1,166 @@
+        }
+
+        setupObservers()
+
+        viewModel.start()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        outState.putSerializable(WordPress.SITE, viewModel.site)
+        super.onSaveInstanceState(outState)
+    }
+
+    private fun setupObservers() {
+        viewModel.events.observe(this, Observer<List<ActivityLogModel>> {
+            reloadEvents()
+        })
+
+        viewModel.eventListStatus.observe(this, Observer<ActivityLogViewModel.ActivityLogListStatus> { listStatus ->
+            refreshProgressBars(listStatus)
+        })
+    }
+
+    private fun refreshProgressBars(eventListStatus: ActivityLogViewModel.ActivityLogListStatus?) {
+        if (!isAdded || view == null) {
+            return
+        }
+        // We want to show the swipe refresher for the initial fetch but not while loading more
+        swipeToRefreshHelper.isRefreshing = eventListStatus === ActivityLogViewModel.ActivityLogListStatus.FETCHING
+        // We want to show the progress bar at the bottom while loading more but not for initial fetch
+        val showLoadMore = eventListStatus === ActivityLogViewModel.ActivityLogListStatus.LOADING_MORE
+        activityLogListProgress.visibility = if (showLoadMore) View.VISIBLE else View.GONE
+    }
+
+    private fun reloadEvents() {
+        setEvents(viewModel.events.value ?: emptyList())
+    }
+
+    private fun setEvents(events: List<ActivityLogModel>) {
+        context?.let {
+            val adapter: ActivityLogAdapter
+            if (activityLogList.adapter == null) {
+                adapter = ActivityLogAdapter(it)
+                activityLogList.adapter = adapter
+            } else {
+                adapter = activityLogList.adapter as ActivityLogAdapter
+            }
+            adapter.updateList(events)
+        }
+    }
+
+    inner class ActivityLogAdapter(context: Context) : RecyclerView.Adapter<ActivityLogViewHolder>() {
+        private val list = ArrayList<ActivityLogModel>()
+        private var layoutInflater: LayoutInflater = LayoutInflater.from(context)
+
+        override fun onBindViewHolder(holder: ActivityLogViewHolder, position: Int) {
+            val item = getItem(position)
+            holder.bind(item)
+
+            if (position == itemCount - 1) {
+                viewModel.loadMore()
+            }
+
+            holder.header.visibility = if (shouldDisplayHeader(position)) View.VISIBLE else View.GONE
+            holder.button.visibility = if (item.rewindable == true) View.VISIBLE else View.GONE
+        }
+
+        init {
+            setHasStableIds(true)
+        }
+
+        internal fun updateList(items: List<ActivityLogModel>) {
+            val diffResult = DiffUtil.calculateDiff(ActivityLogDiffCallback(list, items))
+            list.clear()
+            list.addAll(items)
+            diffResult.dispatchUpdatesTo(this)
+        }
+
+        private fun shouldDisplayHeader(position: Int): Boolean {
```

Comment:
I've created a new `ActivityLogListItemViewModel` that contains only the necessary properties that are already formatted. But for checking whether to display the header you need to look at the previous item, so that logic can't be moved.